### DISCUSSION TITLE
How to attach a correlation id (mdc) to all log messages in the request scope?

### DISCUSSION THREAD

**Discussion by u6f6o** [klogging/klogging#264] [Q&A] [answered]
Hi, 

I am currently migrating an existing application from quarkus to spring web flux with coroutines. In the process I also learned about klogging and gave it a try. I quite like it so far but was wondering how I could use mdc like mechanics with klogging, meaning, adding a correlation-id property automatically to every log message in the current request scope?

Best regards, 
u6f6o

**Comment by mjstrasser** [comment, marked as answer]
Hi @u6f6o, thanks for trying Klogging!

Are correlation IDs in the Reactor context already? 

If so, you can configure Spring to copy specified keys into the Klogging context. Here is an example configuration that will copy the values of keys `correlationId`, `tracingId` and `spanId` from the Reactor context to the Klogging context and will appear in logs.

```kotlin
package com.example.config

import io.klogging.context.Context
import io.klogging.events.EventItems
import kotlinx.coroutines.reactor.ReactorContext
import org.springframework.context.annotation.Configuration

const val correlationKey = "correlationId"
const val tracingKey = "tracingId"
const val spanKey = "spanId"

@Configuration
class CopyReactorToKloggingContext {
    init {
        val copyContextItems: (ReactorContext) -> EventItems = { reactorContext ->
            buildMap {
                listOf(correlationKey, tracingKey, spanKey).forEach { key ->
                    if (reactorContext.context.hasKey(key))
                        put(key, reactorContext.context[key])
                }
            }
        }
        Context.addContextItemExtractor(ReactorContext, copyContextItems)
    }
}
```

(I haven’t used Spring in a while so this might not be the most effective way to configure it, but it works.)

If you don’t have correlation IDs in request scope already, you can try adding Micrometer tracing to your app to generate `traceId` values. 

Otherwise you could get a `X-Correlation-ID` value from an HTTP header or generate a new one, and add it to the Reactor context. Here is some sample code:

```kotlin
package com.example.filters

import com.example.config.correlationKey
import org.springframework.stereotype.Component
import org.springframework.web.server.ServerWebExchange
import org.springframework.web.server.WebFilter
import org.springframework.web.server.WebFilterChain
import reactor.core.publisher.Mono
import java.util.UUID

const val headerKey = "X-Correlation-ID"

@Component
class CorrelationIdFilter : WebFilter {
    override fun filter(exchange: ServerWebExchange, chain: WebFilterChain): Mono<Void> {
        val headers = exchange.request.headers
        val correlationId: String = headers[headerKey]?.first() ?: UUID.randomUUID().toString()
        return chain
            .filter(exchange)
            .contextWrite { ctx -> ctx.put(correlationKey, correlationId) }
    }
}
```

**Comment by u6f6o** [comment]
@mjstrasser ty very much for your detailed explanation!
